### ISSUE TITLE
Suggested JSON formatting

### ISSUE THREAD

**Issue by jedwards1211** [jedwards1211/metacave#16]
@vpicaver that's a good point in #15 about version control history, it would suck not to be able to read diffs at all.

I guess the spec could recommend that programs insert newlines between station and shot objects in metacave files?
That would at least make survey data diffs fairly readable without much extra bytes.

We could also recommend that caves and trip objects have one property per line.  Or at least that all properties except a cave's trips be on one line, and all properties except a trip's survey be on one line.

I wish this were more trivial to implement with JSON output tools that I know, though it could be worse.

**Comment by vpicaver** [comment]
I'll probably be using [QJsonDocument::IndentFormat](https://doc.qt.io/qt-5/qjsondocument.html#JsonFormat-enum). I think it'll add newlines to every property or array entry by default. We should also recommend that properties are in alphabetical order.  I think QJsonDocument does orders properties by default if it's populated with QVariantMap. 

I know there's several git diff plugins for diffing json files:
https://gist.github.com/oscherler/22eb60eb78a4e4e89e0506578e8b820a
https://github.com/benjamine/jsondiffpatch
https://github.com/andreyvit/json-diff
https://t-a-w.blogspot.com/2016/05/sensible-git-diff-for-json-files.html

**Comment by vpicaver** [comment]
Here's one for merging json files in git:
https://openbase.io/js/git-json-merge#:~:text=git%2Djson%2Dmerge-,A%20git%20merge%20driver%20that%20use%20xdiff%20to%20automatically%20resolve,by%20git%2Dpo%2Dmerge.

**Comment by jedwards1211** [comment]
Having newlines and indentation on every property blows up the file size big time though :) for example shot measurement
 properties are at like 8 or 9 levels of indentation so that's at least that many bytes extra per measurement.

**Comment by vpicaver** [comment]
Yea, the indentions will take more data than the data it self in some places. I don't think it'll be that big of an issue. I guess it could double the file size. At least it'll compresses well. Currently the CaveWhere files are super massive > 1GB for 1 miles of cave, because of how the survey scans are put into the file. I'm hoping break the CaveWhere file up into smaller chunks. 

Perhaps we could develop a metacave reformatter library that takes a json string and reformats it correctly to balance newlines and indents. And it would also reduce git conflicts between applications.

**Comment by jedwards1211** [comment]
Yeah I forgot how git (and any version control worth its salt I guess) stores the diffs compressed, so that's good...
I do remember being kinda shocked how much bigger the file was with full indentation (I want to say more than double using two space indents) but I'll have to go back and experiment with it.

I wonder if some of those JSON merging tools can preserve formatting where things haven't changed?  If I could make my progs preserve the input formatting of all the JSON that wasn't changed, I think that would be a nice benefit for working with version control.
